src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,403 @@
+    // else. Note that the class cannot be final since it's extended by the
+    // internal versions. If someone will define a constructor with any other
+    // visibility, then anyone will be able to extend the class, which is not what
+    // we want.
+    HybridTopFieldDocSortCollector(final int numHits, final HitsThresholdChecker hitsThresholdChecker, final Sort sort) {
+        this.numHits = numHits;
+        this.hitsThresholdChecker = hitsThresholdChecker;
+        this.sort = sort;
+    }
+
+    /**
+     * HybridCollectorManager fetches the topDocs in the reduce method.
+     * @return List of TopFieldDocs which represents results of Top Docs of individual subquery.
+     */
+    public List<TopFieldDocs> topDocs() {
+        if (compoundScores == null) {
+            return new ArrayList<>();
+        }
+
+        List<TopFieldDocs> topFieldDocs = new ArrayList<>();
+        for (int subQueryNumber = 0; subQueryNumber < compoundScores.length; subQueryNumber++) {
+            topFieldDocs.add(
+                topDocsPerQuery(
+                    0,
+                    Math.min(collectedHits[subQueryNumber], compoundScores[subQueryNumber].size()),
+                    compoundScores[subQueryNumber],
+                    collectedHits[subQueryNumber],
+                    sort.getSort()
+                )
+            );
+        }
+        return topFieldDocs;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return hitsThresholdChecker.scoreMode();
+    }
+
+    protected abstract class HybridTopDocSortLeafCollector implements LeafCollector {
+        protected HybridQueryScorer compoundQueryScorer;
+        private boolean collectedAllCompetitiveHits = false;
+        @Nullable
+        private FieldDoc after;
+        private final Sort sort;
+        private boolean initializePerSegment;
+
+        public HybridTopDocSortLeafCollector(Sort sort, @Nullable FieldDoc after) {
+            this.sort = sort;
+            this.after = after;
+            this.initializePerSegment = true;
+        }
+
+        @Override
+        public void setScorer(final Scorable scorer) throws IOException {
+            if (scorer instanceof HybridQueryScorer) {
+                log.debug("passed scorer is of type HybridQueryScorer, saving it for collecting documents and scores");
+                compoundQueryScorer = (HybridQueryScorer) scorer;
+            } else {
+                compoundQueryScorer = getHybridQueryScorer(scorer);
+                if (Objects.isNull(compoundQueryScorer)) {
+                    log.error(
+                        String.format(Locale.ROOT, "cannot find scorer of type HybridQueryScorer in a hierarchy of scorer %s", scorer)
+                    );
+                }
+            }
+        }
+
+        private HybridQueryScorer getHybridQueryScorer(final Scorable scorer) throws IOException {
+            if (scorer == null) {
```

Comment:
this has to moved to parent function if scorer is null !! I mean in above function.

Reply:
This is the change in the design of how it is done in the current HybridQuery collector. If any change we seek in the Hybrid Query then it can be catered in other issue.

Reply:
@Vikasht34 income scorer object can be a compound  object that has multi level hierarchy of nested scorers. Some of them can be null, and this is a legit case, for example some scenarios of knn query.
We are looking for one scorer that is of type HybridQueryScorer, once that found we stop the search and return that scorer as global result of all recursive calls.
If in the end we couldn't find hybrid scorer we fail with exception, that's why that check is needed in the top level calling function.